### DISCUSSION TITLE
Please please help. I have reached an issue that is preventing me from continuing with the course. I really need your help with this Brownie issue. Timestamp: 4:30:04

### DISCUSSION THREAD

**Discussion by agad98** [smartcontractkit/full-blockchain-solidity-course-py#339] [Q&A] [answered]
I reached Lesson 5 in this course, the part where Patrick installs Brownie.

**I can't for the life of me install Brownie completely through Pipx, Septuptools, or anything. Because my python command doesn't seem to work as normal.**

Preface: So I followed the course exactly to the letter up to this point. I installed everything that I was told to. Web3, Ganache-Cli, everything...

Come to the Brownie part and all hell breaks loose....

First, when I run the command `python3 -m pip install --user pipx`
I get
`python3 : The term python3 is not recognized as the name of a cmdlet, 
function, script file, or operable program. Check the spelling of the       
name, or if a path was included, verify that the path is correct and try    
again.`

Also with `python3 -m pipx ensurepath`
I get `python3 : The term python3 is not recognized as the name of a cmdlet, 
function, script file, or operable program. Check the spelling of the       
name, or if a path was included, verify that the path is correct and try    
again.`

However, If I run `python -m pip install --user pipx` it work and I get:

```
WARNING: Ignoring invalid distribution -ip (c:\python310\lib\site-packages)
WARNING: Ignoring invalid distribution - (c:\python310\lib\site-packages)   
WARNING: Ignoring invalid distribution -ip (c:\python310\lib\site-packages) 
WARNING: Ignoring invalid distribution - (c:\python310\lib\site-packages)   
Requirement already satisfied: pipx in c:\users\ahmed\appdata\roaming\python\python310\site-packages (0.16.4)
Requirement already satisfied: argcomplete<2.0,>=1.9.4 in c:\users\ahmed\appdata\roaming\python\python310\site-packages (from pipx) (1.12.3)
Requirement already satisfied: userpath>=1.6.0 in c:\users\ahmed\appdata\roaming\python\python310\site-packages (from pipx) (1.7.0)
Requirement already satisfied: colorama>=0.4.4 in c:\users\ahmed\appdata\roaming\python\python310\site-packages (from pipx) (0.4.4)
Requirement already satisfied: packaging>=20.0 in c:\users\ahmed\appdata\roaming\python\python310\site-packages (from pipx) (21.2)
Requirement already satisfied: pyparsing<3,>=2.0.2 in c:\users\ahmed\appdata\roaming\python\python310\site-packages (from packaging>=20.0->pipx) (2.4.7)Requirement already satisfied: click in c:\users\ahmed\appdata\roaming\python\python310\site-packages (from userpath>=1.6.0->pipx) (8.0.3)
WARNING: Ignoring invalid distribution -ip (c:\python310\lib\site-packages)
WARNING: Ignoring invalid distribution - (c:\python310\lib\site-packages)   
WARNING: Ignoring invalid distribution -ip (c:\python310\lib\site-packages)
WARNING: Ignoring invalid distribution - (c:\python310\lib\site-packages)
```

And If I run `python -m pipx ensurepath` I get:
```
C:\Users\Ahmed\.local\bin is already in PATH.
C:\Users\Ahmed\AppData\Roaming\Python\Python310\Scripts has been been       
    added to PATH, but you need to open a new terminal or re-login for      
    this PATH change to take effect.

You will need to open a new terminal or re-login for the PATH changes to    
take effect.

Otherwise pipx is ready to go! ✨ 🌟 ✨
```

Cool.. so then I try running `pipx install eth-brownie` it should work right? No. it doesn't. I get:
```
pipx : The term 'pipx' is not recognized as the name of a cmdlet, 
function, script file, or operable program. Check the spelling of the       
name, or if a path was included, verify that the path is correct and try    
again.
At line:1 char:1
+ pipx install eth-brownie
+ ~~~~
    + CategoryInfo          : ObjectNotFound: (pipx:String) [], CommandNot  
   FoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

If I do `python pipx install eth-brownie` I get:
`C:\Python310\python.exe: can't open file 'C:\\Users\\Ahmed\\Documents\\demos\\brownie_simple_storage\\pipx': [Errno 2] No such file or directory  `

If I do: `py -m pipx install eth-brownie` it works initially and I get 
`⡿ installing eth-brownie`

but then I get the following error right after:
```
Fatal error from pip prevented installation. Full pip output in file:
    C:\Users\Ahmed\.local\pipx\logs\cmd_2021-11-10_14.17.53_pip_errors.log  

pip failed to build package:
    cytoolz

Some possibly relevant errors from pip install:
    build\lib.win-amd64-3.10\cytoolz\functoolz.cp310-win_amd64.pyd : fatal error LNK1120: 1 unresolved externals
    error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x64\\link.exe' failed with exit code 1120

Error installing eth-brownie.
```

**Here are the solutions I have attempted so far:**

- I added the Python location to PATH
- I added the Scripts folder location to PATH
- I deleted and re-installed Python **FROM THEIR WEBSITE**
- I deleted and re-installed MS VS Code
- I deleted and re-installed the Python Extension
- I added .local to the Path
- I added PYTHONPATH to System Variables with the addresses of both Python and Scripts folder

_**Why isn't my Python command working? Why isn't my PIPX working even though it is clearly installed?**_

I do not know what to do anymore. This has been very frustrating and annoying, causing me to stop the course and troubleshoot. I have spent about 7 hours so far trying to fix it with no avail.

**Comment by agad98** [comment]
Also this issue happens with Yarn. I can install it normally using:
`npm install --global yarn`
```
changed 1 package, and audited 2 packages in 518ms        

found 0 vulnerabilities
```

But I check for it, and it is not there:
```
yarn : The term 'yarn' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try 
again.
At line:1 char:1
+ yarn --version
+ ~~~~
    + CategoryInfo          : ObjectNotFound: (yarn:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```
**There seems to be a disrepency between where things are downloaded and where they are getting called from. **

**Comment by cromewar** [comment, marked as answer]
Hello @agad98, Oof you have a lot of problems, I'm going to try to resolve all of them.

1. Regarding python and python3, this depends a lot on your operating system, Unix based systems comes with python installed, sometimes with both 2.7 and 3.x versions, so that's why Patrick uses python3, as you are on windows just use python and don't worry about it.
2. Windows is not good enough to recognize packages installed with `pip` (basically it doesn't add them to environment variables). so you'll have to use `python -m` every time.
3. Recommendation: don't use pipx, trust me you'll have problems with vscode if you install eth-brownie using pipx... use old trusty `pip` instead.
4. regarding to the error at installing eth-brownie (cytoolz), this is a tricky one.. means you need some `C` libraries for windows that don't come included with de OS, so there are two possible paths here:
   - Install Visual Studio Community and the Python development tools (is heavy and is going to take a while) 
   - Install WSL (windows subsystem for linux) and use linux for your python stuff. (faster but you'll need to learn some basic linux commands)
5. same as pip packages, windows does not recognize packages installed with `NPM`, you should fix this adding them to your Path, follow [this](https://stackoverflow.com/questions/27864040/fixing-npm-path-in-windows-8-and-10) thread.

My personal recommendation: use WSL, as I said you would have to spend some time getting along with Linux console, but is totally worth it... you will handle all your development libraries like a pro (python and npm included) and you wont mess up with your windows config. If you want a guide about how to setup windows for this use [this](https://medium.com/@cromewar/how-to-setup-windows-10-11-for-smart-contract-development-and-brownie-e7d8d13555b3) guide.

Best regards, Cromewar.

**Reply by agad98** [reply]
Thank you!

**Reply by toonboifresh** [reply]
I second using WSL, this isn't the first time I've struggled with Windows/python and all of the other misc packages.  Watch out for WSL's virtualization requirements, that's a whole other hurdle to get over but it's well worth it.

If you are having issues with virtualization pre-reqs - try disabling virtualization/WSL/hypervisor under windows features, reboot, and then enable everything required, and reboot again.  Also, make sure your BIOS is configured appropriately as well for virtualization.

It sounds crazy, but I spent hours trying to get WSL to install only to find that all I had to do was disable the windows features required and re-enable them.

**Reply by cromewar** [reply]
Thanks for the tip @toonboifresh .
